src/main/python/mmlspark/cyber/anomaly/collaborative_filtering.py

```diff
@@ -0,0 +1,880 @@
+
+import numpy as np
+
+from pyspark.ml import Estimator, Transformer
+from pyspark.ml.param.shared import Param, Params
+from pyspark.ml.recommendation import ALS
+from pyspark.sql import DataFrame, functions as f, types as t
+
+"""
+Glossary:
+the term 'res' in this package is a shorthand for resource
+"""
+
+
+def _make_dot():
+    """
+    create a method that performs a dot product between two vectors (list of doubles)
+    :return: the method
+    """
+    @f.udf(t.DoubleType())
+    def dot(v, u):
+        if (v is not None) and (u is not None):
+            vv = np.pad(np.array(v), (0, len(u) - len(v)), 'constant', constant_values=1.0) if len(v) < len(
+                u) else np.array(v)
+            uu = np.pad(np.array(u), (0, len(v) - len(u)), 'constant', constant_values=1.0) if len(u) < len(
+                v) else np.array(u)
+
+            return float(vv.dot(uu))
+        else:
+            return None
+
+    return dot
+
+
+class CfAlgoParams:
+    """
+    Parameter for the AccessAnomaly Estimator indicating
+    if to use implicit or explicit feedback versions of the ALS algorithm
+    and additional parameters that are relevant for the implicit/explicit versions
+    """
+    def __init__(self, implicit: bool):
+        self._implicit = implicit
+        self._alpha: Optional[float] = None
+        self._complementset_factor: Optional[int] = None
+        self._neg_score: Optional[float] = None
+
+        # set default values
+        if implicit:
+            self.set_alpha(1.0)
+        else:
+            self.set_complementset_factor(2)
+            self.set_neg_score(1.0)
+
+    @property
+    def implicit(self) -> bool:
+        return self._implicit
+
+    @property
+    def alpha(self) -> Optional[float]:
+        """
+        Relevant for the implicit version only
+        :return: alpha value (see descriptions in explainParam of AccessAnomaly)
+        """
+        return self._alpha
+
+    @property
+    def complementset_factor(self) -> Optional[int]:
+        """
+        Relevant for the explicit version only
+        :return: complementset_factor value (see descriptions in explainParam of AccessAnomaly)
+        """
+        return self._complementset_factor
+
+    @property
+    def neg_score(self) -> Optional[float]:
+        """
+        Relevant for the explicit version only
+        :return: neg_score value (see descriptions in explainParam of AccessAnomaly)
+        """
+        return self._neg_score
+
+    def set_alpha(self, value: float):
+        assert self.implicit
+        self._alpha = value
+        return self
+
+    def set_complementset_factor(self, value: int):
+        assert not self.implicit
+        self._complementset_factor = value
+        return self
+
+    def set_neg_score(self, value: float):
+        assert not self.implicit
+        assert value > 0.0
+        self._neg_score = value
+        return self
+
+
+class AccessAnomalyConfig:
```

Comment:
These should be parameters on the estimator and configurable instead of static

Reply:
Discussed and agreed this is fine